### PR TITLE
Add option "enable_pxe"

### PR DESCRIPTION
**Pull request by toabctl**
This option is enabled by default which means, that the dhcp-server
answers to requests from yet unknown hosts.

### CODE REVIEW

File: chef/cookbooks/provisioner/recipes/dhcp_update.rb [hunk 1]
```diff
@@ -4,9 +4,12 @@
...
 admin_net = node[:network][:networks]["admin"]
 lease_time = node[:provisioner][:dhcp]["lease-time"]
-pool_opts = {
-  "dhcp" => ['allow unknown-clients',
-             '      if option arch = 00:06 {
+
+pool_opts = {}
+if node[:provisioner][:dhcp]["for-unknown-nodes"]
+
+pool_opts["dhcp"] = ['allow unknown-clients',
```
Comment (vuntz) [inline chef/cookbooks/provisioner/recipes/dhcp_update.rb]: Wrong indentation.

Comment (vuntz) [inline chef/cookbooks/provisioner/recipes/dhcp_update.rb]: (also: unneeded empty line)

Comment (toabctl) [inline chef/cookbooks/provisioner/recipes/dhcp_update.rb]: @vuntz: both fixed in https://github.com/toabctl/barclamp-provisioner/commit/4cc31e30cfab616a3a1938db92d523f0db7ce0b7

File: chef/data_bags/crowbar/bc-template-provisioner.json [hunk 2]
```diff
@@ -75,6 +75,7 @@
       "use_serial_console": false,
       "serial_tty": "ttyS0,115200n8",
       "dhcp": {
+        "for-unknown-nodes": true,
```
Comment (vuntz) [inline chef/data_bags/crowbar/bc-template-provisioner.json]: Apologies, I didn't see that yesterday, but we're avoiding adding new attributes with `-`, and always use `_` instead now.

File: chef/data_bags/crowbar/bc-template-provisioner.json [hunk 3]
```diff
@@ -101,7 +102,7 @@
     "provisioner": {
       "crowbar-revision": 0,
       "crowbar-applied": false,
-      "schema-revision": 12,
+      "schema-revision": 13,
```
Comment (vuntz) [inline chef/data_bags/crowbar/bc-template-provisioner.json]: Do we need to bump to 20, or did we already bump to 10 for tex?

**Comment by vuntz** [comment]
Btw, I think I'm guessing the use case, but it might be worth describing it in the commit message.

**Comment by bmwiedemann** [comment]
https://trello.com/c/yPPs5sy2/275-3-p12-disable-dhcp-pxe-boot-dynamic-range-in-crowbar-want-to-only-use-crowbar-register ... just discussed with Max that this might also be useful for mkcloud-testing of crowbar-register

File: chef/data_bags/crowbar/bc-template-provisioner.json (hunk 3 above)
Comment (toabctl) [inline chef/data_bags/crowbar/bc-template-provisioner.json]: Bump needed. Done in https://github.com/toabctl/barclamp-provisioner/commit/822fab6c191a4c559ebc2975a9a8f82403b57753

File: chef/data_bags/crowbar/bc-template-provisioner.json (hunk 2 above)
Comment (toabctl) [inline chef/data_bags/crowbar/bc-template-provisioner.json]: Done in https://github.com/toabctl/barclamp-provisioner/commit/822fab6c191a4c559ebc2975a9a8f82403b57753

**Comment by toabctl** [comment]
> Btw, I think I'm guessing the use case, but it might be worth describing it in the commit message.

Also done in https://github.com/toabctl/barclamp-provisioner/commit/822fab6c191a4c559ebc2975a9a8f82403b57753 .

**Comment by toabctl** [comment]
I'll check disableing the tftpd server should be done in this PR as well.

**Comment by aspiers** [comment]
The last sentence of the commit message has a typo and is also not quite clear.  Please can you change it to:

> Disabling this option is useful for setups where nodes are only deployed via "crowbar_register" and never via sledgehammer.

File: chef/cookbooks/provisioner/recipes/dhcp_update.rb [hunk 4]
```diff
@@ -4,9 +4,11 @@
 admin_ip = Chef::Recipe::Barclamp::Inventory.get_network_by_type(node, "admin").address
 admin_net = node[:network][:networks]["admin"]
 lease_time = node[:provisioner][:dhcp]["lease-time"]
-pool_opts = {
-  "dhcp" => ['allow unknown-clients',
-             '      if option arch = 00:06 {
+
+pool_opts = {}
+if node[:provisioner][:dhcp]["for_unknown_nodes"]
```
Comment (tboerger) [inline chef/cookbooks/provisioner/recipes/dhcp_update.rb]: Why are you mixing symbol keys with string keys? Just use one of them!

**Comment by toabctl** [comment]
This is still WIP and not really tested. I'll test in the next days but wanted to push the code anyway if somebody else takes over the task.

**Comment by tboerger** [comment]
Currently it looks not bad :)

**Comment by toabctl** [comment]
> The last sentence of the commit message has a typo and is also not quite clear. Please can you change it to:
> 
> Disabling this option is useful for setups where nodes are only deployed via "crowbar_register" and never via sledgehammer.

Fixed in https://github.com/toabctl/barclamp-provisioner/commit/0ebac676d64f75db8b1110349c8ac7893be3fcea

**Comment by toabctl** [comment]
Tested and I think it's ready for review now.

File: chef/cookbooks/provisioner/recipes/setup_base_images.rb [hunk 5]
```diff
@@ -209,14 +209,23 @@
...
+    command "pkill in.tftpd"
+    not_if { node[:provisioner][:enable_pxe] }
+    returns [0, 1]
   end
 
   service "xinetd" do
-    running true
-    enabled true
-    action [ :enable, :start ]
+    running node[:provisioner][:enable_pxe] ? true : false
```
Comment (rhafer) [inline chef/cookbooks/provisioner/recipes/setup_base_images.rb]: I know you didn't introduce this. But AFAICS the ```running``` attribute does not really have a effect on the ```service``` resource (it's not even documented). Should we use the chance to clean this up?

File: chef/cookbooks/dhcp/recipes/default.rb [hunk 6]
```diff
@@ -164,6 +164,6 @@
     end
   end
   supports :restart => true, :status => true, :reload => true
-  action :enable
+  action node[:provisioner][:enable_pxe] ? ["enable", "start"] : ["disable", "stop"]
```
Comment (rhafer) [inline chef/cookbooks/dhcp/recipes/default.rb]: I assume we could skip large parts of ```provisioner::update_nodes.rb``` and ```dhcp::default.rb``` did you already investigate that? It might be worth do that in the separate PR after we branched Tex, if it's too heavy for the Tex.

File: chef/cookbooks/provisioner/recipes/setup_base_images.rb (hunk 5 above)
Comment (toabctl) [inline chef/cookbooks/provisioner/recipes/setup_base_images.rb]: > I know you didn't introduce this. But AFAICS the running attribute does not really have a effect on the service resource (it's not even documented). Should we use the chance to clean this up?

"running" is used in some other places. I would like todo that in an extra commit after branching.

Comment (rhafer) [inline chef/cookbooks/provisioner/recipes/setup_base_images.rb]: > "running" is used in some other places. I would like todo that in an extra commit after branching.

Fine with me.

File: chef/cookbooks/dhcp/recipes/default.rb (hunk 6 above)
Comment (toabctl) [inline chef/cookbooks/dhcp/recipes/default.rb]: > I assume we could skip large parts of provisioner::update_nodes.rb and dhcp::default.rb did you already investigate that? It might be worth do that in the separate PR after we branched Tex, if it's too heavy for the Tex.

I disabled ```provisioner::dhcp_update.rb``` and ```provisioner::update_nodes.rb``` when PXE is disabled. 
```dhcp::default.rb``` is a bit more intrusive because this enables/disables the service at the end of the recipe. I skipped that for Tex. Can be done after branching but imho isn't really important.

Comment (toabctl) [inline chef/cookbooks/dhcp/recipes/default.rb]: After a mumble chat with @rhafer , ```provisioner::dhcp_update.rb``` and ```provisioner::update_nodes.rb``` are enabled again. Should be changed in the next release.

Comment (rhafer) [inline chef/cookbooks/dhcp/recipes/default.rb]: > After a mumble chat with @rhafer , provisioner::dhcp_update.rb and provisioner::update_nodes.rb are enabled again. Should be changed in the next release.

Just for completeness: The reason for keeping the update_nodes recipe running for now is, that it also does somethings that are needed when dhcp/pxe is disabled. E.g. when "forggetting" a node it is responsible for deleteing the chef node and role objects of the nodes. I rather not touch this before branching Tex.

**Comment by rhafer** [comment]
+1

File: chef/cookbooks/provisioner/recipes/setup_base_images.rb [hunk 7]
```diff
@@ -209,14 +209,23 @@
...
-    action [ :enable ]
+    enabled node[:provisioner][:enable_pxe] ? true : false
+    action node[:provisioner][:enable_pxe] ? "enable" : "disable"
+  end
+
+  # NOTE(toabctl): stop for tftp does not really help. the process gets started
+  # by xinetd and has a default timeout of 900 seconds which triggers when no
+  # new connections start in this period. So kill the process here
+  execute "kill in.tftpd process" do
+    command "pkill in.tftpd"
```
Comment (vuntz) [inline chef/cookbooks/provisioner/recipes/setup_base_images.rb]: I'm not a big fan of having a `pkill` every 15 minutes, but can't find a better alternative.

**Comment by vuntz** [comment]
+1

File: chef/cookbooks/provisioner/recipes/setup_base_images.rb (hunk 7 above)
Comment (toabctl) [inline chef/cookbooks/provisioner/recipes/setup_base_images.rb]: > I'm not a big fan of having a pkill every 15 minutes, but can't find a better alternative.

We could also decrease the timeout of the daemon and block the traffic with a iptables rule. But I prefer the ```pkill``` version.
